src/mon/OSDMonitor.cc

```diff
@@ -886,7 +886,11 @@ void OSDMonitor::prime_pg_temp(
 			    &next_acting, &next_acting_primary);
   if (acting == next_acting)
     return;  // no change since last epoch
-
+  if (acting.size() > next_acting.size()) {
+    // do not touch if acting size will shrink
+    // otherwise we'll probably add some non-existent osds into pg_temp
+    return;
+  }
```

Comment:
I'm worried this will skip some cases where a primed pg_temp helps. Can we instead filter acting to remove !exists osds?  It seems possible there are some corner cases that this missed (e.g., if the osd is destroyed in the same epoch that some other CRUSH change is made).